src/ice_nucleation.F90

```diff
@@ -0,0 +1,563 @@
+    type(env_state_t), intent(in) :: env_state_initial
+    !> Environment state at the end of the timestep. The rel_humid
+    !> value will be ignored and overwritten with a new value.
+    type(env_state_t), intent(inout) :: env_state_final
+    !> Total time to integrate.
+    real(kind=dp), intent(in) :: del_t
+    !> Immersion freezing scheme type (e.g., 1:constant rate, 2:sigular, 3:ABIFM).
+    integer, intent(in) :: immersion_freezing_scheme_type
+    !> Freezing rate (only used for the constant rate scheme).
+    real(kind=dp) :: freezing_rate
+
+
+    !> Call the immersion freezing subroutine according to the immersion
+    !> freezing scheme.
+    if ((immersion_freezing_scheme_type .eq. IMMERSION_FREEZING_SCHEME_ABIFM) &
+       .OR. (immersion_freezing_scheme_type .eq. IMMERSION_FREEZING_SCHEME_CONST)) then
+       if (do_speedup) then
+          call immersion_freezing_time_dependent(aero_state, aero_data, env_state_initial, &
+            env_state_final, del_t, immersion_freezing_scheme_type, freezing_rate)
+       else
+          call immersion_freezing_time_dependent_naive(aero_state, aero_data, env_state_initial, &
+            env_state_final, del_t, immersion_freezing_scheme_type, freezing_rate)
+       end if
+    else if (immersion_freezing_scheme_type .eq. &
+        IMMERSION_FREEZING_SCHEME_SINGULAR) then
+       call immersion_freezing_singular(aero_state, aero_data, env_state_initial, &
+          env_state_final)
+    else
+       call assert_msg(121370299, .false., &
+            'Error type of immersion freezing scheme')
+    endif
+
+  end subroutine immersion_freezing
+
+!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+
+  !> Initialization for the sigular scheme, sampling the freezing temperature
+  !> for each particles.
+  subroutine singular_initialize(aero_state, aero_data)
+    implicit none
+    !> Aerosol state.
+    type(aero_state_t), intent(inout) :: aero_state
+    !> Aerosol data.
+    type(aero_data_t), intent(in) :: aero_data
+
+    integer :: i_part
+    real(kind=dp) :: a_INAS, b_INAS, p, S, T0, temp
+    real(kind=dp) :: aerosol_diameter
+
+    T0 = const%water_freeze_temp
+    do i_part = 1, aero_state_n_part(aero_state)
+       a_INAS = -0.517 
+       b_INAS = 8.934
+       aerosol_diameter = aero_particle_dry_diameter(aero_state%apa%particle(i_part), aero_data)
+       S = const%pi * aerosol_diameter **2
+       p = pmc_random()
+       temp = (log(1 - p) + exp(-S * exp(-a_INAS * T0 + b_INAS))) / (-S)
+       aero_state%apa%particle(i_part)%imf_temperature = T0 + (log(temp) &
+          - b_INAS) / a_INAS
+    end do
+  end subroutine singular_initialize
+
+!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+
+  !> Simulation for singular scheme, deciding whether to freeze for each
+  !> particle. Run in each time step.
+  subroutine immersion_freezing_singular(aero_state, aero_data, env_state_initial, env_state_final)
```

Comment:
`env_state_initial` is passed here and never used (and if this is correct, `env_state_final` can just become `env_state`).